[Bug] [Xlang] Overloaded Java Class can't be called via xlang

### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.


### Ray Component

Others

### What happened + What you expected to happen

Given the following Java interface and class
```java
public interface ExampleInterface {
  default String echo(String inp) {
    return inp;
  }
}
```

```java
public class ExampleImpl implements ExampleInterface {
  @Override
  public String echo(String inp) {
    return inp + " echo";
  }
}
```

and calling it via Python (use own package of course):
```python
import ray
ray.init(job_config=ray.job_config.JobConfig(code_search_path=["."]))
cls = ray.java_actor_class("io.ray.serve.util.ExampleImpl")
handle = cls.remote()
print(ray.get(handle.echo.remote("hi")))
```

I get the following error:
```
In [6]: ray.get(handle.echo.remote("hi"))                                                                    
---------------------------------------------------------------------------
CrossLanguageError                        Traceback (most recent call last)
<ipython-input-6-eb72b94dc879> in <module>
----> 1 ray.get(handle.echo.remote("hi"))

~/Desktop/ray/ray/python/ray/_private/client_mode_hook.py in wrapper(*args, **kwargs)
    103             if func.__name__ != "init" or is_client_mode_enabled_by_default:
    104                 return getattr(ray, func.__name__)(*args, **kwargs)
--> 105         return func(*args, **kwargs)
    106 
    107     return wrapper

~/Desktop/ray/ray/python/ray/worker.py in get(object_refs, timeout)
   1742                     raise value.as_instanceof_cause()
   1743                 else:
-> 1744                     raise value
   1745 
   1746         if is_individual_id:

CrossLanguageError: An exception raised from JAVA:
io.ray.runtime.exception.RayTaskException: (pid=68724, ip=192.168.1.69) Error executing task 63964fa4841d4a2eeb3d013876f4f3c8c527dbba01000000
        at io.ray.runtime.task.TaskExecutor.execute(TaskExecutor.java:188)
        at io.ray.runtime.RayNativeRuntime.nativeRunTaskExecutor(Native Method)
        at io.ray.runtime.RayNativeRuntime.run(RayNativeRuntime.java:251)
        at io.ray.runtime.runner.worker.DefaultWorker.main(DefaultWorker.java:15)
Caused by: java.lang.RuntimeException: RayFunction io.ray.serve.util.ExampleImpl.echo is overloaded, the signature can't be empty.
        at io.ray.runtime.functionmanager.FunctionManager$JobFunctionTable.getFunction(FunctionManager.java:183)
        at io.ray.runtime.functionmanager.FunctionManager.getFunction(FunctionManager.java:108)
        at io.ray.runtime.task.TaskExecutor.getRayFunction(TaskExecutor.java:69)
        at io.ray.runtime.task.TaskExecutor.execute(TaskExecutor.java:121)
        ... 3 more
```

Note that if I can make the method not overloaded, then it's passes. 
```
  default String echo(String inp) {
```
to
```
  String echo(String inp);
```

### Versions / Dependencies

master

### Reproduction script

see above

### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

--- a/java/runtime/src/main/java/io/ray/runtime/functionmanager/FunctionManager.java
+++ b/java/runtime/src/main/java/io/ray/runtime/functionmanager/FunctionManager.java
@@ -211,17 +211,9 @@ public class FunctionManager {
           clz = clz.getSuperclass();
         }
 
-        for (Executable e : executables) {
-          LOGGER.info("111=========methodName={}", e.getName());
-        }
-
         // Put interface methods ahead, so that in can be override by subclass methods in `map.put`
         for (Class baseInterface : clazz.getInterfaces()) {
           for (Method method : baseInterface.getDeclaredMethods()) {
-            LOGGER.info(
-                "222=================methodName={}, is_default={}",
-                method.getName(),
-                method.isDefault());
             if (method.isDefault()) {
               executables.add(method);
             }
@@ -239,31 +231,16 @@ public class FunctionManager {
               new RayFunction(
                   e, classLoader, new JavaFunctionDescriptor(className, methodName, signature));
           final boolean isDefault = e instanceof Method && ((Method) e).isDefault();
-          LOGGER.info(
-              "=============className={}, methodname={}, signature={}",
-              className,
-              methodName,
-              signature);
           map.put(
               ImmutablePair.of(methodName, signature), ImmutablePair.of(rayFunction, isDefault));
           // For cross language call java function without signature
           final Pair<String, String> emptyDescriptor = ImmutablePair.of(methodName, "");
           /// default method is not overloaded, so we should filter it.
           if (map.containsKey(emptyDescriptor) && !map.get(emptyDescriptor).getRight()) {
-            LOGGER.info(
-                "Mark this as overloaded =============className={}, methodname={}, signature={}",
-                className,
-                methodName,
-                signature);
             map.put(
                 emptyDescriptor,
                 ImmutablePair.of(null, false)); // Mark this function as overloaded.
           } else {
-            LOGGER.info(
-                "Not Mark this as overloaded =============className={}, methodname={}, signature={}",
-                className,
-                methodName,
-                signature);
             map.put(emptyDescriptor, ImmutablePair.of(rayFunction, isDefault));
           }
         }

--- a/java/runtime/src/main/java/io/ray/runtime/functionmanager/FunctionManager.java
+++ b/java/runtime/src/main/java/io/ray/runtime/functionmanager/FunctionManager.java
@@ -1,5 +1,6 @@
 package io.ray.runtime.functionmanager;
 
+import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Lists;
 import io.ray.api.function.RayFunc;
 import io.ray.api.id.JobId;
@@ -157,7 +158,7 @@ public class FunctionManager {
     /** The job's corresponding class loader. */
     final ClassLoader classLoader;
     /** Functions per class, per function name + type descriptor. */
-    ConcurrentMap<String, Map<Pair<String, String>, RayFunction>> functions;
+    ConcurrentMap<String, Map<Pair<String, String>, Pair<RayFunction, Boolean>>> functions;
 
     JobFunctionTable(ClassLoader classLoader) {
       this.classLoader = classLoader;
@@ -165,7 +166,7 @@ public class FunctionManager {
     }
 
     RayFunction getFunction(JavaFunctionDescriptor descriptor) {
-      Map<Pair<String, String>, RayFunction> classFunctions = functions.get(descriptor.className);
+      Map<Pair<String, String>, Pair<RayFunction, Boolean>> classFunctions = functions.get(descriptor.className);
       if (classFunctions == null) {
         synchronized (this) {
           classFunctions = functions.get(descriptor.className);
@@ -176,7 +177,7 @@ public class FunctionManager {
         }
       }
       final Pair<String, String> key = ImmutablePair.of(descriptor.name, descriptor.signature);
-      RayFunction func = classFunctions.get(key);
+      RayFunction func = classFunctions.get(key).getLeft();
       if (func == null) {
         if (classFunctions.containsKey(key)) {
           throw new RuntimeException(
@@ -192,9 +193,11 @@ public class FunctionManager {
     }
 
     /** Load all functions from a class. */
-    Map<Pair<String, String>, RayFunction> loadFunctionsForClass(String className) {
+    Map<Pair<String, String>, Pair<RayFunction, Boolean>> loadFunctionsForClass(String className) {
       // If RayFunction is null, the function is overloaded.
-      Map<Pair<String, String>, RayFunction> map = new HashMap<>();
+      // The value of this map is a pair of <rayFunction, isDefault>.
+      // The `isDefault` is used to mark if the method is a marked as default keyword.
+      Map<Pair<String, String>, Pair<RayFunction, Boolean>> map = new HashMap<>();
       try {
         Class clazz = Class.forName(className, true, classLoader);
         List<Executable> executables = new ArrayList<>();
@@ -208,9 +211,14 @@ public class FunctionManager {
           clz = clz.getSuperclass();
         }
 
+        for (Executable e : executables) {
+          LOGGER.info("111=========methodName={}", e.getName());
+        }
+
         // Put interface methods ahead, so that in can be override by subclass methods in `map.put`
         for (Class baseInterface : clazz.getInterfaces()) {
           for (Method method : baseInterface.getDeclaredMethods()) {
+            LOGGER.info("222=================methodName={}, is_default={}", method.getName(), method.isDefault());
             if (method.isDefault()) {
               executables.add(method);
             }
@@ -227,13 +235,18 @@ public class FunctionManager {
           RayFunction rayFunction =
               new RayFunction(
                   e, classLoader, new JavaFunctionDescriptor(className, methodName, signature));
-          map.put(ImmutablePair.of(methodName, signature), rayFunction);
+          final boolean isDefault = e instanceof Method && ((Method) e).isDefault();
+          LOGGER.info("=============className={}, methodname={}, signature={}", className, methodName, signature);
+          map.put(ImmutablePair.of(methodName, signature), ImmutablePair.of(rayFunction, isDefault));
           // For cross language call java function without signature
           final Pair<String, String> emptyDescriptor = ImmutablePair.of(methodName, "");
-          if (map.containsKey(emptyDescriptor)) {
-            map.put(emptyDescriptor, null); // Mark this function as overloaded.
+          /// default method is not overloaded, so we should filter it.
+          if (map.containsKey(emptyDescriptor) && !map.get(emptyDescriptor).getRight()) {
+            LOGGER.info("Mark this as overloaded =============className={}, methodname={}, signature={}", className, methodName, signature);
+            map.put(emptyDescriptor, ImmutablePair.of(null, false)); // Mark this function as overloaded.
           } else {
-            map.put(emptyDescriptor, rayFunction);
+            LOGGER.info("Not Mark this as overloaded =============className={}, methodname={}, signature={}", className, methodName, signature);
+            map.put(emptyDescriptor, ImmutablePair.of(rayFunction, isDefault));
           }
         }
       } catch (Exception e) {

--- a/java/runtime/src/main/java/io/ray/runtime/functionmanager/FunctionManager.java
+++ b/java/runtime/src/main/java/io/ray/runtime/functionmanager/FunctionManager.java
@@ -1,6 +1,5 @@
 package io.ray.runtime.functionmanager;
 
-import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Lists;
 import io.ray.api.function.RayFunc;
 import io.ray.api.id.JobId;
@@ -166,7 +165,8 @@ public class FunctionManager {
     }
 
     RayFunction getFunction(JavaFunctionDescriptor descriptor) {
-      Map<Pair<String, String>, Pair<RayFunction, Boolean>> classFunctions = functions.get(descriptor.className);
+      Map<Pair<String, String>, Pair<RayFunction, Boolean>> classFunctions =
+          functions.get(descriptor.className);
       if (classFunctions == null) {
         synchronized (this) {
           classFunctions = functions.get(descriptor.className);
@@ -218,7 +218,10 @@ public class FunctionManager {
         // Put interface methods ahead, so that in can be override by subclass methods in `map.put`
         for (Class baseInterface : clazz.getInterfaces()) {
           for (Method method : baseInterface.getDeclaredMethods()) {
-            LOGGER.info("222=================methodName={}, is_default={}", method.getName(), method.isDefault());
+            LOGGER.info(
+                "222=================methodName={}, is_default={}",
+                method.getName(),
+                method.isDefault());
             if (method.isDefault()) {
               executables.add(method);
             }
@@ -236,16 +239,31 @@ public class FunctionManager {
               new RayFunction(
                   e, classLoader, new JavaFunctionDescriptor(className, methodName, signature));
           final boolean isDefault = e instanceof Method && ((Method) e).isDefault();
-          LOGGER.info("=============className={}, methodname={}, signature={}", className, methodName, signature);
-          map.put(ImmutablePair.of(methodName, signature), ImmutablePair.of(rayFunction, isDefault));
+          LOGGER.info(
+              "=============className={}, methodname={}, signature={}",
+              className,
+              methodName,
+              signature);
+          map.put(
+              ImmutablePair.of(methodName, signature), ImmutablePair.of(rayFunction, isDefault));
           // For cross language call java function without signature
           final Pair<String, String> emptyDescriptor = ImmutablePair.of(methodName, "");
           /// default method is not overloaded, so we should filter it.
           if (map.containsKey(emptyDescriptor) && !map.get(emptyDescriptor).getRight()) {
-            LOGGER.info("Mark this as overloaded =============className={}, methodname={}, signature={}", className, methodName, signature);
-            map.put(emptyDescriptor, ImmutablePair.of(null, false)); // Mark this function as overloaded.
+            LOGGER.info(
+                "Mark this as overloaded =============className={}, methodname={}, signature={}",
+                className,
+                methodName,
+                signature);
+            map.put(
+                emptyDescriptor,
+                ImmutablePair.of(null, false)); // Mark this function as overloaded.
           } else {
-            LOGGER.info("Not Mark this as overloaded =============className={}, methodname={}, signature={}", className, methodName, signature);
+            LOGGER.info(
+                "Not Mark this as overloaded =============className={}, methodname={}, signature={}",
+                className,
+                methodName,
+                signature);
             map.put(emptyDescriptor, ImmutablePair.of(rayFunction, isDefault));
           }
         }

--- a/java/runtime/src/test/java/io/ray/runtime/functionmanager/FunctionManagerTest.java
+++ b/java/runtime/src/test/java/io/ray/runtime/functionmanager/FunctionManagerTest.java
@@ -183,7 +183,7 @@ public class FunctionManagerTest {
   @Test
   public void testLoadFunctionTableForClass() {
     JobFunctionTable functionTable = new JobFunctionTable(getClass().getClassLoader());
-    Map<Pair<String, String>, RayFunction> res =
+    Map<Pair<String, String>, Pair<RayFunction, Boolean>> res =
         functionTable.loadFunctionsForClass(ChildClass.class.getName());
     // The result should be 4 entries:
     //   1, the constructor with signature
@@ -211,7 +211,7 @@ public class FunctionManagerTest {
                 overloadFunctionDescriptorDouble.signature)));
     Assert.assertTrue(res.containsKey(ImmutablePair.of(overloadFunctionDescriptorInt.name, "")));
     Pair<String, String> overloadKey = ImmutablePair.of(overloadFunctionDescriptorInt.name, "");
-    RayFunction func = res.get(overloadKey);
+    RayFunction func = res.get(overloadKey).getLeft();
     // The function is overloaded.
     Assert.assertTrue(res.containsKey(overloadKey));
     Assert.assertNull(func);

--- a/java/test/src/main/java/io/ray/test/CrossLanguageInvocationTest.java
+++ b/java/test/src/main/java/io/ray/test/CrossLanguageInvocationTest.java
@@ -396,4 +396,13 @@ public class CrossLanguageInvocationTest extends BaseTest {
 
     private byte[] value;
   }
+
+  public void testPyCallJavaOeveridedMethodWithDefault() {
+    ObjectRef<Object> res =
+      Ray.task(PyFunction.of(PYTHON_MODULE, "py_func_call_java_overrided_method_with_default_keyword", Object.class))
+        .remote();
+    System.out.println(res.get());
+    Assert.assertEquals("hi".getBytes(StandardCharsets.UTF_8), res.get());
+  }
+
 }

--- a/java/test/src/main/java/io/ray/test/CrossLanguageInvocationTest.java
+++ b/java/test/src/main/java/io/ray/test/CrossLanguageInvocationTest.java
@@ -399,10 +399,13 @@ public class CrossLanguageInvocationTest extends BaseTest {
 
   public void testPyCallJavaOeveridedMethodWithDefault() {
     ObjectRef<Object> res =
-      Ray.task(PyFunction.of(PYTHON_MODULE, "py_func_call_java_overrided_method_with_default_keyword", Object.class))
-        .remote();
+        Ray.task(
+                PyFunction.of(
+                    PYTHON_MODULE,
+                    "py_func_call_java_overrided_method_with_default_keyword",
+                    Object.class))
+            .remote();
     System.out.println(res.get());
     Assert.assertEquals("hi".getBytes(StandardCharsets.UTF_8), res.get());
   }
-
 }

--- a/java/test/src/main/java/io/ray/test/CrossLanguageInvocationTest.java
+++ b/java/test/src/main/java/io/ray/test/CrossLanguageInvocationTest.java
@@ -405,7 +405,6 @@ public class CrossLanguageInvocationTest extends BaseTest {
                     "py_func_call_java_overrided_method_with_default_keyword",
                     Object.class))
             .remote();
-    System.out.println(res.get());
-    Assert.assertEquals("hi".getBytes(StandardCharsets.UTF_8), res.get());
+    Assert.assertEquals("hi", res.get());
   }
 }

--- a/java/test/src/main/java/io/ray/test/ExampleImpl.java
+++ b/java/test/src/main/java/io/ray/test/ExampleImpl.java
@@ -1,0 +1,10 @@
+package io.ray.test;
+
+public class ExampleImpl implements ExampleInterface {
+
+  @Override
+  public String echo(String str) {
+    return str;
+  }
+
+}

--- a/java/test/src/main/java/io/ray/test/ExampleImpl.java
+++ b/java/test/src/main/java/io/ray/test/ExampleImpl.java
@@ -6,5 +6,4 @@ public class ExampleImpl implements ExampleInterface {
   public String echo(String str) {
     return str;
   }
-
 }

--- a/java/test/src/main/java/io/ray/test/ExampleInterface.java
+++ b/java/test/src/main/java/io/ray/test/ExampleInterface.java
@@ -1,0 +1,9 @@
+package io.ray.test;
+
+public interface ExampleInterface {
+
+   default String echo(String str) {
+     return "default" + str;
+   }
+
+}

--- a/java/test/src/main/java/io/ray/test/ExampleInterface.java
+++ b/java/test/src/main/java/io/ray/test/ExampleInterface.java
@@ -2,8 +2,7 @@ package io.ray.test;
 
 public interface ExampleInterface {
 
-   default String echo(String str) {
-     return "default" + str;
-   }
-
+  default String echo(String str) {
+    return "default" + str;
+  }
 }

--- a/java/test/src/main/resources/test_cross_language_invocation.py
+++ b/java/test/src/main/resources/test_cross_language_invocation.py
@@ -133,8 +133,7 @@ def py_func_get_and_invoke_named_actor():
     return b"true"
 
 @ray.remote
-def py_func_f():
+def py_func_call_java_overrided_method_with_default_keyword():
     cls = ray.java_actor_class("io.ray.test.ExampleImpl")
     handle = cls.remote()
-    print(ray.get(handle.echo.remote("hi")))
-    return b"ok"
+    return ray.get(handle.echo.remote("hi"))

--- a/java/test/src/main/resources/test_cross_language_invocation.py
+++ b/java/test/src/main/resources/test_cross_language_invocation.py
@@ -132,6 +132,7 @@ def py_func_get_and_invoke_named_actor():
     assert ray.get(java_named_actor.concat.remote(b"world")) == b"helloworld"
     return b"true"
 
+
 @ray.remote
 def py_func_call_java_overrided_method_with_default_keyword():
     cls = ray.java_actor_class("io.ray.test.ExampleImpl")

--- a/java/test/src/main/resources/test_cross_language_invocation.py
+++ b/java/test/src/main/resources/test_cross_language_invocation.py
@@ -131,3 +131,10 @@ def py_func_get_and_invoke_named_actor():
     java_named_actor = ray.get_actor("java_named_actor")
     assert ray.get(java_named_actor.concat.remote(b"world")) == b"helloworld"
     return b"true"
+
+@ray.remote
+def py_func_f():
+    cls = ray.java_actor_class("io.ray.test.ExampleImpl")
+    handle = cls.remote()
+    print(ray.get(handle.echo.remote("hi")))
+    return b"ok"

